include/seqan3/core/parallel/detail/latch.hpp

```diff
@@ -0,0 +1,166 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::latch.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <atomic>
+#include <cassert>
+
+#include <seqan3/core/parallel/detail/spin_delay.hpp>
+#include <seqan3/std/new>
+
+namespace seqan3::detail
+{
+
+/*!\brief A single-use synchronisation point to coordinate concurrent threads.
+ * \ingroup parallel
+ *
+ * \details
+ *
+ * A latch is a thread coordination mechanism that allows any number of threads to block until an expected
+ * count is summed (exactly) by threads that arrived at the latch. The expected count is set when the latch is
+ * constructed. An individual latch is a single-use object; once the count has been reached, the latch cannot be reused.
+ * This implementation uses a lock-free mechanism if the atomic operations on the respective platform are lock-free.
+ *
+ * \note This adapts the proposal for the c++ standard
+ * [P0666R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0666r2.pdf) for latches and is likely to be
+ * changed according to the revisions of the proposal.
+ */
+class latch
+{
+public:
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    latch()                           = delete;  //!< Deleted.
+    latch(latch const &)              = delete;  //!< Deleted.
+    latch(latch &&)                   = default; //!< Defaulted.
+    latch & operator=(latch const &)  = delete;  //!< Deleted.
+    latch & operator=(latch &&)       = default; //!< Defaulted.
+
+    //!\brief Destructs the latch and waits for all participating threads to arrive.
+    ~latch()
+    {
+        spin_delay delay{};
+        while (num_waiting.load(std::memory_order_acquire) > 0)
+            delay.wait();
+    }
+
+    /*!\brief Constructs the latch with the expected number of threads.
+     * \param expected The number of threads participating in this synchronisation point.
+     */
+    explicit latch(ptrdiff_t const expected) :
+        counter{expected}
+    {
+        assert(expected >= 0);
+        num_waiting.store(0, std::memory_order_relaxed);
```

Comment:
You use `std::memory_order_relaxed` instead of the default? This was from the proposal, right?